django/contrib/messages/storage/cookie.py

```diff
@@ -150,16 +153,23 @@ def _decode(self, data):
         """
         if not data:
             return None
-        bits = data.split('$', 1)
-        if len(bits) == 2:
-            hash, value = bits
-            if constant_time_compare(hash, self._hash(value)):
-                try:
-                    # If we get here (and the JSON decode works), everything is
-                    # good. In any other case, drop back and return None.
-                    return json.loads(value, cls=MessageDecoder)
-                except json.JSONDecodeError:
-                    pass
+        bits = data.split('$', 2)
+        if len(bits) >= 2:
+            # Before Django 3.1, no algorithm prefix was set.
+            hash_algorithm = bits[0] if len(bits) == 3 else 'sha1'
+            _hash, value = bits[-2:]
+            try:
+                equals = constant_time_compare(_hash, self._hash(value, algorithm=hash_algorithm))
+            except InvalidAlgorithm:
+                pass
+            else:
+                if equals:
+                    try:
+                        # If we get here (and the JSON decode works), everything is
+                        # good. In any other case, drop back and return None.
+                        return json.loads(value, cls=MessageDecoder)
+                    except json.JSONDecodeError:
+                        pass
```

Comment:
I think we should add release notes that old format is deprecated and will not be valid in Django 4.0+.